### PR TITLE
[dag] reliable broadcast receiver

### PR DESCRIPTION
**Pull request by ibalajiarun**
### Description

This PR adds the trait and structs for the receiver-side of the reliable broadcast. 

### Test Plan

added unit tests

### CODE REVIEW

File: consensus/src/dag/types.rs [hunk 1]
```diff
@@ -298,3 +290,172 @@ pub struct DAGNetworkMessage {
     #[serde(with = "serde_bytes")]
     pub data: Vec<u8>,
 }
+
+#[derive(Clone, Serialize, Deserialize, Debug)]
+pub enum DAGMessage {
```
Comment (zekun000) [inline consensus/src/dag/types.rs:295]: so many boilerplate 🤦 but I don't see a simpler way to do it better, maybe we can do a procmacro but not today

File: consensus/src/network.rs [hunk 2]
```diff
@@ -39,11 +40,30 @@ use futures::{
     stream::{select, select_all},
     SinkExt, Stream, StreamExt,
 };
+use serde::{de::DeserializeOwned, Serialize};
 use std::{
     mem::{discriminant, Discriminant},
     time::Duration,
 };
 
+pub trait ConsensusMessageTrait: Sized + Clone + Serialize + DeserializeOwned {
```
Comment (zekun000) [inline consensus/src/network.rs:49]: i personally prefer shorter names, there's conventions in other languages to name it like TConsensusMessage, TDagMessage

File: consensus/src/dag/dag_network.rs [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+    fn process(&mut self, message: Self::Message) -> anyhow::Result<Self::Ack>;
+}
+
+#[async_trait]
+pub trait DAGNetworkSender: Send + Sync {
+    async fn send_rpc(
+        &self,
+        receiver: Author,
+        message: DAGMessage,
```
Comment (zekun000) [inline consensus/src/dag/dag_network.rs:18]: does it need to be DAGMessage? since we may want to move it to general consensus

File: consensus/src/dag/reliable_broadcast.rs [hunk 4]
```diff
@@ -78,3 +71,46 @@ impl ReliableBroadcast {
         }
     }
 }
+
+pub struct NodeBroadcastHandler {
+    signatures_by_round_peer: BTreeMap<Round, BTreeMap<Author, NodeDigestSignature>>,
```
Comment (zekun000) [inline consensus/src/dag/reliable_broadcast.rs:76]: do we need to store the Node here?

**Review by sasha8 — COMMENTED** [review]
A few comments.

File: consensus/src/dag/types.rs [hunk 5]
```diff
@@ -83,12 +94,48 @@ impl NodeMetadata {
     }
 }
 
+#[derive(Clone, Serialize, Deserialize, CryptoHasher, Debug)]
+pub struct NodeCertificateMetadata {
```
Comment (zekun000) [inline consensus/src/dag/types.rs:98]: I'd just call it NodeCertificate, like quorum certificate, it can contain some metadata without metadata in the name 😄

File: consensus/src/dag/reliable_broadcast.rs [hunk 6]
```diff
@@ -67,3 +84,108 @@ impl ReliableBroadcast {
...
+            signer,
+            verifier,
+        }
+    }
+
+    pub fn gc_before_round(&mut self, min_round: Round) {
+        self.signatures_by_round_peer.retain(|r, _| r >= &min_round);
+    }
+
+    fn validate(&self, node: &Node) -> anyhow::Result<()> {
```
Comment (zekun000) [inline consensus/src/dag/reliable_broadcast.rs:123]: we should add a `fn verify(self, verifier: &ValidatorVerifier) -> Result<Self>` to TDagMessage for all self contained verification

File: consensus/src/dag/reliable_broadcast.rs [hunk 7]
```diff
@@ -67,3 +84,108 @@ impl ReliableBroadcast {
...
+    type Request = Node;
+    type Response = NodeDigestSignature;
+
+    fn process(&mut self, node: Self::Request) -> anyhow::Result<Self::Response> {
+        self.validate(&node)?;
+
+        let signatures_by_peer = self
+            .signatures_by_round_peer
+            .entry(node.metadata().round())
+            .or_insert(BTreeMap::new());
```
Comment (zekun000) [inline consensus/src/dag/reliable_broadcast.rs:179]: need a TODO to persist data

File: consensus/src/dag/dag_handler.rs [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+                    }
+                }
+            }
+        }
+    }
+
+    async fn process_rpc(&mut self, rpc_request: IncomingDAGRequest) -> anyhow::Result<()> {
+        let dag_message: DAGMessage = ConsensusMessageTrait::from_network_message(rpc_request.req)?;
+        let response: anyhow::Result<DAGMessage> = match dag_message {
+            DAGMessage::NodeMsg(node) => self.node_receiver.process(node).map(|r| r.into()),
```
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply consensus/src/dag/dag_handler.rs:49]: I will follow up in the next PR.

**Comment by ibalajiarun** [comment]
I updated the PR. It's getting too long, so I will defer additional validation tests to the next PR.

File: consensus/src/dag/dag_handler.rs (hunk 8 above)
[the remark this answers is not in the document]
Comment (ibalajiarun) [reply consensus/src/dag/dag_handler.rs:49]: I will spin it into a separate task after adding additional more handlers.

**Comment by ibalajiarun** [comment]
Okay, I have added the validation unit tests too.

File: consensus/src/dag/dag_fetcher.rs [hunk 9]
```diff
@@ -110,19 +73,20 @@ impl DagFetcher {
...
                 .await
-                .and_then(FetchResponse::from_network_message)
+                .and_then(DAGMessage::try_from)
+                .and_then(FetchResponse::try_from)
                 .and_then(|response| response.verify(&request, &self.epoch_state.verifier))
             {
                 // TODO: support chunk response or fallback to state sync
                 let mut dag_writer = self.dag.write();
-                for rounds in response.certifies_nodes {
+                for rounds in response.certified_nodes() {
```
Comment (zekun000) [inline consensus/src/dag/dag_fetcher.rs:87]: nit: we should just consume this to avoid clone

File: consensus/src/dag/reliable_broadcast.rs [hunk 10]
```diff
@@ -67,3 +80,118 @@ impl ReliableBroadcast {
...
+            .filter(|parent| !dag_reader.exists(parent.metadata().digest()))
+            .cloned()
+            .collect();
+        if !missing_parents.is_empty() {
+            // For each missing parent, verify their signatures and voting power
+            ensure!(
+                missing_parents.iter().all(|parent| {
+                    let node_digest = NodeDigest::new(*parent.metadata().digest());
+                    self.verifier
+                        .verify_multi_signatures(&node_digest, parent.signatures())
```
Comment (zekun000) [inline consensus/src/dag/reliable_broadcast.rs:150]: I think this checks voting power inside, I'd slightly prefer we do all the self-contained checks (including parents certificate check) for node outside but let's do that in another pr.

File: consensus/src/dag/types.rs [hunk 11]
```diff
@@ -164,38 +184,86 @@ impl Node {
...
+        let prev_round = current_round - 1;
+        // check if the parents' round is the node's round - 1
+        ensure!(
+            self.parents()
+                .iter()
+                .all(|parent| parent.metadata().round() == prev_round),
+            "invalid parent round"
+        );
+
+        ensure!(
```
Comment (zekun000) [inline consensus/src/dag/types.rs:214]: need to handle round 0?

**Review by zekun000 — APPROVED** [review]

File: consensus/src/dag/types.rs (hunk 11 above)
Comment (ibalajiarun) [reply consensus/src/dag/types.rs:214]: it is a special case I am handling in `reliable_broadcast.rs`

**Review by msmouse — APPROVED** [review]
I'm blind.

File: consensus/src/dag/reliable_broadcast.rs [hunk 12]
```diff
@@ -67,3 +80,107 @@ impl ReliableBroadcast {
...
+        ensure!(
+            prev_round >= dag_reader.lowest_round(),
+            NodeBroadcastHandleError::MissingParents
+        );
+
+        // check which parents are missing in the DAG
+        let missing_parents: Vec<NodeCertificate> = node
+            .parents()
+            .iter()
+            .filter(|parent| !dag_reader.exists(parent.metadata().digest()))
```
Comment (sasha8) [inline consensus/src/dag/reliable_broadcast.rs:141]: We need to compare metadata.

**Review by sasha8 — COMMENTED** [review]
One TODO we discussed offline.

File: consensus/src/dag/reliable_broadcast.rs (hunk 12 above)
Comment (zekun000) [reply consensus/src/dag/reliable_broadcast.rs:141]: instead of this, I think we should calculate the digest either in signing or verification

Comment (sasha8) [reply consensus/src/dag/reliable_broadcast.rs:141]: Balaji was saying we don't have enough info to compute parents digests.
